integration/stability/soak_parallel_rm.sh

```diff
@@ -0,0 +1,168 @@
+# then just set this to a very large number
+MAX_CONTAINERS=110
+
+function count_containers() {
+	docker ps -qa | wc -l
+}
+
+function check_all_running() {
+	echo Checking ${how_many} containers have all relevant components
+
+	# check what docker thinks
+	how_many_running=$(count_containers)
+
+	if (( ${how_many_running} != ${how_many} )); then
+		echo "Wrong number of containers running (${how_many_running} != ${how_many}) - stopping"
+		exit -1
+	fi
+
+	# check we have the right number of proxy's
+	how_many_proxys=$(ps --no-header -C ${PROXY_NAME} | wc -l)
+	if (( ${how_many_running} >= 1 )); then
+		# If we have any containers, then we expect to have a single proxy
+		if (( ${how_many_proxys} != 1 )); then
+			echo "Wrong number of proxys running (${how_many_running} containers, ${how_many_proxys} proxys) - stopping"
+			exit -1
+		fi
+	else
+		# No containers running, but if we have run before then the proxy may still be running
+		# It is not clear if the proxy quits if there are no containers left
+		if (( ${how_many_proxys} > 1 )); then
+			echo "Wrong number of proxys running (${how_many_running} containers, ${how_many_proxys} proxys) - stopping"
+			exit -1
+		fi
+	fi
+
+	# check we have the right number of shims
+	how_many_shims=$(ps --no-header -C ${SHIM_NAME} | wc -l)
+	# two shim processes per container...
+	if (( ${how_many_running}*2 != ${how_many_shims} )); then
+		echo "Wrong number of shims running (${how_many_running}*2 != ${how_many_shims} - stopping)"
+		exit -1
+	fi
+
+	# check we have the right number of qemu's
+	how_many_qemus=$(ps --no-header -C ${QEMU_NAME} | wc -l)
+	if (( ${how_many_running} != ${how_many_qemus} )); then
+		echo "Wrong number of qemus running (${how_many_running} != ${how_many_qemus} - stopping)"
+		exit -1
+	fi
+
+	# check we have no runtimes running (they should be transient, we should not 'see them')
+	how_many_runtimes=$(ps --no-header -C ${RUNTIME_NAME} | wc -l)
+	if (( ${how_many_runtimes} )); then
+		echo "Wrong number of runtimes running (${how_many_runtimes} - stopping)"
+		exit -1
+	fi
+}
+
+# reported system 'available' memory
+function get_system_avail() {
+	echo $(free -b | head -2 | tail -1 | awk '{print $7}')
+}
+
+function go() {
+	echo "Running..."
+
+	how_many=0
+
+	while true; do {
+		check_all_running
+
+		echo "Run $RUNTIME: $PAYLOAD: $COMMAND"
+		docker run --runtime=${RUNTIME} -tid ${PAYLOAD} ${COMMAND}
+
+		((how_many++))
+		if (( ${how_many} > ${MAX_CONTAINERS} )); then
+			echo "And we have hit the max ${how_many} containers"
+			return
+		fi
+
+		how_much=$(get_system_avail)
+		if (( ${how_much} < ${MEM_CUTOFF} )); then
+			echo "And we are out of memory on container ${how_many} (${how_much} < ${MEM_CUTOFF})"
+			return
+		fi
+	}
+	done
+}
+
+function kill_all_containers() {
+	present=$(docker ps -qa | wc -l)
+	if ((${present})); then
+		docker rm -f $(docker ps -qa)
+	fi
+}
+
+function init() {
+	kill_all_containers
```

Comment:
could be good to restart docker/cc-proxy services at this point

Reply:
Hmm, yea, I guess. It almost seem backwards - make the system 'more stable' to try and 'find instabilities', but for repeatability it makes a lot of sense. Let me see if I can do something by re-using the stuff we have in the metrics lib files (would be good if we could share some of that across metrics and QA for instance).

Reply:
maybe it is time to add a `lib` directory in the root of the repository, and start populating it and using it. But this can be another PR.

Reply:
An idea from @chavafg - we can also check that there are no pending docker mounts left after the test - something else we have spotted as a potential issue (https://github.com/clearcontainers/packaging/pull/110)